File: webboard/src/model/merge.test.ts
import { describe, expect, it } from 'vitest';
import type { GraphEdge, GraphNode, GraphSnapshot } from './graph';
import { mergeSnapshots } from './merge';

function node(id: string, kind: GraphNode['kind'], extra: Partial<GraphNode> = {}): GraphNode {
  return { id, type: id, displayName: id.split('.').pop() ?? id, kind, boundTo: [], ...extra };
}

function edge(from: string, to: string): GraphEdge {
  return { id: `${from} -> ${to} # 0`, from, to, siteKind: 'constructorParam', deferred: 'none' };
}

function snap(partial: Partial<GraphSnapshot>): GraphSnapshot {
  return {
    schemaVersion: 2,
    appId: 'com.example.app',
    variant: 'debug',
    scopes: [],
    nodes: [],
    edges: [],
    ...partial,
  };
}

describe('mergeSnapshots (multi-module fragments)', () => {
  it('replaces a consumer fragment external placeholder with the owning module node', () => {
    // :app consumed core.UserRepository — its fragment only knows it as external.
    const app = snap({
      nodes: [node('app.Presenter', 'injectable'), node('core.UserRepository', 'external')],
      edges: [edge('app.Presenter', 'core.UserRepository')],
    });
    // :core owns the interface — a boundInterface satellite of the implementation.
    const core = snap({
      appId: 'unknown',
      nodes: [
        node('core.NetworkUserRepository', 'injectable', { boundTo: ['core.UserRepository'] }),
        node('core.UserRepository', 'boundInterface'),
      ],
    });

    const merged = mergeSnapshots(app, core === undefined ? [] : [core]);
    expect(merged.nodes.find((n) => n.id === 'core.UserRepository')?.kind).toBe('boundInterface');
    expect(merged.nodes).toHaveLength(3);
    expect(merged.appId).toBe('com.example.app'); // primary wins
  });

  it('keeps the real node no matter the fragment order', () => {
    const withReal = snap({ nodes: [node('core.Analytics', 'injectable')] });
    const withPlaceholder = snap({ nodes: [node('core.Analytics', 'external')] });
    for (const fragments of [[withReal, withPlaceholder], [withPlaceholder, withReal]]) {
      const merged = mergeSnapshots(snap({}), fragments);
      expect(merged.nodes).toHaveLength(1);
      expect(merged.nodes[0]!.kind).toBe('injectable');
    }
  });

  it('unions edges and scopes without duplicates, sorted', () => {
    const a = snap({
      scopes: [{ name: 'Singleton', level: 0 }, { name: 'Session', level: 10 }],
      nodes: [node('a.A', 'injectable'), node('b.B', 'external')],
      edges: [edge('a.A', 'b.B')],
    });
    const b = snap({
      scopes: [{ name: 'Singleton', level: 0 }, { name: 'ActivityScoped', level: 1 }],
      nodes: [node('b.B', 'injectable')],
      edges: [edge('a.A', 'b.B')],
    });
    const merged = mergeSnapshots(a, [b]);
    expect(merged.scopes.map((s) => s.name)).toEqual(['Singleton', 'ActivityScoped', 'Session']);
    expect(merged.edges).toHaveLength(1);
    expect(merged.nodes.map((n) => n.id)).toEqual(['a.A', 'b.B']);
  });

  it('is the identity for a single-module app', () => {
    const app = snap({
      nodes: [node('a.A', 'injectable')],
      edges: [],
      scopes: [{ name: 'Singleton', level: 0 }],
    });
    expect(mergeSnapshots(app, [])).toEqual(app);
  });
});

File: kite/board/src/test/kotlin/com/kite/di/board/MergeTest.kt
package com.kite.di.board

import com.kite.di.graph.GraphEdge
import com.kite.di.graph.GraphNode
import com.kite.di.graph.GraphSnapshot
import com.kite.di.graph.NodeKind
import com.kite.di.graph.ScopeDef
import com.kite.di.graph.SiteKind
import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * Merging the per-module fragments into the one canvas the board draws. The case
 * that matters is the cross-module dependency: the consuming module only knows it
 * as a dangling `external` placeholder, and the owning module knows what it is.
 */
class MergeTest {

    private fun node(id: String, kind: NodeKind = NodeKind.INJECTABLE) =
        GraphNode(id = id, type = id, displayName = id.substringAfterLast('.'), kind = kind)

    private fun edge(id: String, from: String, to: String) =
        GraphEdge(id = id, from = from, to = to, siteKind = SiteKind.CONSTRUCTOR_PARAM)

    private fun snapshot(
        appId: String = "com.example",
        nodes: List<GraphNode> = emptyList(),
        edges: List<GraphEdge> = emptyList(),
        scopes: List<ScopeDef> = emptyList(),
    ) = GraphSnapshot(appId = appId, variant = "debug", nodes = nodes, edges = edges, scopes = scopes)

    @Test
    fun `the owning module's real node replaces a consumer's external placeholder`() {
        val app = snapshot(nodes = listOf(node("core.Analytics", NodeKind.EXTERNAL)))
        val core = snapshot(nodes = listOf(node("core.Analytics", NodeKind.INJECTABLE)))

        val merged = mergeSnapshots(app, listOf(core))

        assertEquals(1, merged.nodes.size)
        assertEquals(NodeKind.INJECTABLE, merged.nodes.single().kind)
    }

    @Test
    fun `a real node is not downgraded by a placeholder from another fragment`() {
        val app = snapshot(nodes = listOf(node("core.Analytics", NodeKind.INJECTABLE)))
        val other = snapshot(nodes = listOf(node("core.Analytics", NodeKind.EXTERNAL)))

        assertEquals(NodeKind.INJECTABLE, mergeSnapshots(app, listOf(other)).nodes.single().kind)
    }

    @Test
    fun `nodes, edges and scopes unite without duplicates and come out sorted`() {
        val app = snapshot(
            nodes = listOf(node("b.B"), node("a.A")),
            edges = listOf(edge("e2", "a.A", "b.B")),
            scopes = listOf(ScopeDef("Singleton", 0)),
        )
        val fragment = snapshot(
            nodes = listOf(node("a.A"), node("c.C")),
            edges = listOf(edge("e2", "a.A", "b.B"), edge("e1", "c.C", "a.A")),
            scopes = listOf(ScopeDef("Singleton", 0), ScopeDef("ActivityScoped", 1)),
        )

        val merged = mergeSnapshots(app, listOf(fragment))

        assertEquals(listOf("a.A", "b.B", "c.C"), merged.nodes.map { it.id })
        assertEquals(listOf("e1", "e2"), merged.edges.map { it.id })
        assertEquals(listOf("Singleton", "ActivityScoped"), merged.scopes.map { it.name })
    }

    @Test
    fun `identity comes from the application fragment, not the libraries`() {
        val app = snapshot(appId = "com.example.app")
        val library = snapshot(appId = "com.example.core")

        assertEquals("com.example.app", mergeSnapshots(app, listOf(library)).appId)
    }

    @Test
    fun `merging is deterministic regardless of fragment order`() {
        val app = snapshot(nodes = listOf(node("a.A")))
        val one = snapshot(nodes = listOf(node("b.B")))
        val two = snapshot(nodes = listOf(node("c.C")))

        assertEquals(
            mergeSnapshots(app, listOf(one, two)).nodes.map { it.id },
            mergeSnapshots(app, listOf(two, one)).nodes.map { it.id },
        )
    }
}
